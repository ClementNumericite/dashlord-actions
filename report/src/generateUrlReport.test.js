// @ts-nocheck

const fs = require("fs");
const generateUrlReport = require("./generateUrlReport");

jest.mock("fs");

const testUrl = "https://www.test.com";
const b64 = "aHR0cHM6Ly93d3cudGVzdC5jb20=";

// mock a json in the latest scan
const mockJson = (name, data) =>
  jest.mock(
    `results/${b64}/9876/${name}`,
    () =>
      data
        ? data
        : {
            data: name,
          },
    { virtual: true }
  );

// for some reason jest.resetModules doesnt work with virtual mocks
const unMockJson = (name) => {
  jest.unmock(`results/${b64}/9876/${name}`);
};

describe("generateUrlReport", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  test(`should return null on invalid url`, () => {
    expect(
      generateUrlReport({
        url: "https://www.invalid.com",
      })
    ).toEqual(null);
  });

  test(`should generate latest report for a valid url`, () => {
    fs.existsSync.mockImplementationOnce(() => true); // check url folder
    fs.existsSync.mockImplementationOnce(() => false); // screenshot
    fs.readdirSync.mockImplementationOnce(() => ["9876", "1234", "5678"]); // assume scans folders use a sortable date-format

    // mock all required files
    mockJson("codescanalerts.json");
    mockJson("dependabotalerts.json");
    mockJson("http.json");
    mockJson("lhr.json");
    mockJson("nmapvuln.json");
    mockJson("nuclei.json", []);
    mockJson("testssl.json");
    mockJson("thirdparties.json");
    mockJson("updownio.json");
    mockJson("wappalyzer.json");
    mockJson("zap.json");

    expect(
      generateUrlReport({
        url: testUrl,
      })
    ).toMatchSnapshot();

    // unmock all required files
    unMockJson("codescanalerts.json");
    unMockJson("dependabotalerts.json");
    unMockJson("http.json");
    unMockJson("lhr.json");
    unMockJson("nmapvuln.json");
    unMockJson("nuclei.json", []);
    unMockJson("testssl.json");
    unMockJson("thirdparties.json");
    unMockJson("updownio.json");
    unMockJson("wappalyzer.json");
    unMockJson("zap.json");
  });
  test(`should allow empty/invalid reports`, () => {
    fs.existsSync.mockImplementationOnce(() => true); // check url folder
    fs.existsSync.mockImplementationOnce(() => false); // screenshot
    fs.readdirSync.mockImplementationOnce(() => ["9876", "1234", "5678"]); // assume scans folders use a sortable date-format

    mockJson("codescanalerts.json");
    mockJson("wappalyzer.json");

    expect(
      generateUrlReport({
        url: testUrl,
      })
    ).toMatchSnapshot();

    unMockJson("codescanalerts.json");
    unMockJson("wappalyzer.json");
  });

  test(`should detect screenshot if any`, () => {
    fs.existsSync.mockImplementationOnce(() => true); // check url folder
    fs.existsSync.mockImplementationOnce(() => true); // screenshot
    fs.readdirSync.mockImplementationOnce(() => ["9876", "1234", "5678"]); // assume scans folders use a sortable date-format
    expect(
      generateUrlReport({
        url: testUrl,
      }).screenshot
    ).toEqual(true);
  });
});
