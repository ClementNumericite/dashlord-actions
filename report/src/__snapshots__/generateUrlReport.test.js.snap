// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateUrlReport should generate latest report for a valid url 1`] = `
Object {
  "codescan": Object {
    "data": "codescanalerts.json",
  },
  "dependabot": Object {
    "data": "dependabotalerts.json",
  },
  "http": Object {
    "data": "http.json",
  },
  "lhr": Object {
    "audits": Object {},
    "categories": Object {},
    "fetchTime": undefined,
    "finalUrl": undefined,
    "requestedUrl": undefined,
    "runWarnings": undefined,
  },
  "nmap": Object {
    "data": "nmapvuln.json",
  },
  "nuclei": Array [],
  "scores": Object {
    "codescan": 0,
    "dependabot": 0,
    "http": 0,
    "lhr": 0,
    "nmap": 0,
    "nuclei": 0,
    "screenshot": 0,
    "testssl": 0,
    "thirdparties": 0,
    "updownio": 0,
    "url": 0,
    "wappalyzer": 0,
    "zap": 0,
  },
  "screenshot": false,
  "testssl": Object {
    "data": "testssl.json",
  },
  "thirdparties": Object {
    "data": "thirdparties.json",
  },
  "updownio": Object {
    "data": "updownio.json",
  },
  "url": "https://www.test.com",
  "wappalyzer": Object {
    "data": "wappalyzer.json",
  },
  "zap": Object {
    "data": "zap.json",
    "site": undefined,
  },
}
`;

exports[`generateUrlReport should skip scans if some not found 1`] = `
Object {
  "codescan": Object {
    "data": "codescanalerts.json",
  },
  "dependabot": null,
  "http": null,
  "lhr": null,
  "nmap": null,
  "nuclei": null,
  "scores": Object {
    "codescan": 0,
    "dependabot": 0,
    "http": 0,
    "lhr": 0,
    "nmap": 0,
    "nuclei": 0,
    "screenshot": 0,
    "testssl": 0,
    "thirdparties": 0,
    "updownio": 0,
    "url": 0,
    "wappalyzer": 0,
    "zap": 0,
  },
  "screenshot": false,
  "testssl": null,
  "thirdparties": null,
  "updownio": null,
  "url": "https://www.test.com",
  "wappalyzer": Object {
    "data": "wappalyzer.json",
  },
  "zap": null,
}
`;
